### ISSUE TITLE
Expose cache TTL configuration / use cache in consecutive runs

### ISSUE THREAD

**Issue by buffcode** [renovatebot/renovate#13798]
### What would you like Renovate to be able to do?

Expose a packageRule configuration on how long to cache certain packages.
Depending on the individual configuration and setup, the cache should be used in consecutive runs.

---

Currently the cache support various adapters, for example Redis. This is used to cache requests / tags / changelog / ....
The default timeout for eg. Redis is 5 minutes, after which the result is discarded [^1] [^2] [^3].

We have a GitLab instance and a single renovate runner. One run over all configured repositories takes around 30-45 minutes.
Most of the projects share common libraries / dependencies such as `node`, `symfony` etc.

Sometimes a host goes down or becomes slow and `abortOnError` kicks in, killing the run for the whole project.
When Renovate starts again after some delay, the whole cache has expired and will likely trigger above error again.

Besides, even without any errors, Renovate will refetch all package information, resulting in slower runs and more traffic.

[^1]: https://github.com/renovatebot/renovate/blob/main/lib/util/cache/package/redis.ts#L38
[^2]: https://github.com/renovatebot/renovate/blob/main/lib/util/cache/package/redis.ts#L64
[^3]: https://github.com/renovatebot/renovate/blob/main/lib/datasource/gitlab-tags/index.ts#L28

### If you have any ideas on how this should be implemented, please tell us here.

```js
{
  "packageRules": [
    // current default
    {
      "matchPackagePatterns": [
        "*"
      ],
      "cache": {
        "ttl": 300 
      }
    },

    // do no re-fetch in the next 24 hours
    {
      "matchPackagePatterns": [
        "node",
        "symfony/*"
      ],
      "cache": {
        "ttl": 86400 
      },
    },

    // always re-fetch / no cache, 
    {
      "matchPackagePatterns": [
        "my-library/that-changes-rapidly"
      ],
      "cache": {
        "ttl": 0 
      }
    }
  ]
}
```

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by rarkins** [comment]
I don't think that cache timeouts should be configurable by users, so adding to packageRules is not desirable.

**Comment by buffcode** [comment]
Currently the cache TTLs (imcomplete list) are set to the following and are IMHO not really consistent / reflect some specific projects or system setup:

| Source | TTL |
| --- | --: |
| GitLab | 55m |
| Docker Labels | 60m |
| Docker Tags | 30m |
| GitHub Release Digest | 1440m |
| GitHub Releases | 10m |
| GitHub Digest | 1440m |
| GitHub Tag Commits | 120m |
| GitHub Tags Digest | 10m |
| GitHub Tags | 10m |
| GitHub Tags Releases | 10m |
| GitHub Tags Digest | 10m |
| GitHub Release Notes | 55m / 1435m / 14495m |
| GitHub Changelog | 55m |

I agree that cache TTLs should not to be subject to the user's configuration but on the other hand: one fixed TTL per Renovate instance is maybe too limiting. Something like `packageRules` (including its `match*` helpers) on a global scope would probably be a better fit.

**Comment by pfyod** [comment]
We have private (maven) packages, that can get updated multiple times per hour, and we would like those packages to get picked up by the renovatebot ASAP. Maven TTL is hardcoded to 24 hours (obviously does not work for our usecase) - we configured renovatebot to bypass maven package cache completely (fake `Authorization` header set via `hostRules` does that).

I think that this ticket is a duplicate of #6573 btw.

**Comment by buffcode** [comment]
@pfyod Thanks for the issue. Though they have similarities, I would like to configure the TTL based on packages, not only by host / manager.

Eg. Symfony updates are quite big (lots of packages) and (in our specific use case) we won't update those packages multiple times a day. On the other hand we have private packages that should be updated more often / asap. Renovate should not spend lots of time updating releases/tags/... and run into othe 

While a host/manager-based approach would be sufficient when those are distinct sources or hosts, it would probably not work for private Packagist/NPM packages hosted on npm.org/packagist itself.

**Comment by rarkins** [comment]
I don't think maven release list ttl is 24 hours

**Comment by pfyod** [comment]
@rarkins: yeah, sorry, it's actually 30 minutes (was reading the wrong place)... Nevertheless, it would make sense (for us) to configure private packages to have different TTL than public.

**Comment by rarkins** [comment]
We then have the choice:
- Configurable TTL per datasource, i.e. ability to override the current hardcoded values
- Configurable TTL per host, which adds some more flexibility for distinguishing between private and public
- Configurable TTL per package. I'm not really in favor of this because I don't think it can be done safely and elegantly

**Comment by pfyod** [comment]
@rarkins: we, for example, proxy maven central via our private repository, so "TTL per host" would probably not have enough flexibility.

**Comment by viceice** [comment]
I would prefer per host configuration, so you can use host rules.

**Comment by rarkins** [comment]
@pfyod at this point we don't consider the increase in complexity necessary to be justified by the benefit, so the most granularity we're willing to support is per-host via hostRules.

**Comment by buffcode** [comment]
+2 for per-package, +1 for per-host.

As long as we don't have a global, immutable, packageRules-like configuration the per-package solution seems to be to massive to implement.
